### ISSUE TITLE
Module is reported as missing, but intellisense works on the module

### ISSUE THREAD

**Issue by daltairwalter** [microsoft/pylance-release#1776]
I have a stub pyi file for a pyd.  Intellisense is working, but I am seeing a warning about a missing file.

In my vscode extensions, I see Pylance v2021.9.0 and Python v2021.9.1191016588

![image](https://user-images.githubusercontent.com/31971208/132401461-6f9f9f2f-bb15-4624-9b5e-4b70a5b7c862.png)

**Comment by jakebailey** [comment]
What is the name of the pyd file? The message says that the source is missing, in this case, we can't find the pyd, but we can find the stub.

**Comment by daltairwalter** [comment]
It wasn't working with my debug pyd which was hstalgocommon_d.pyd.  I tried changing paths to use the release version hstalgocommon.pyd and this is working.

During development, I would prefer not to build release pyd's for this.  Are there any workspace flags I can set to indicate that pylance should look for an _d on the pyds?

**Comment by jakebailey** [comment]
If the path is different, I'm not entirely certain how Python is loading it, since the module name matters. Is this behavior documented somewhere, or is it custom to your project?

You could disable the `reportMissingModuleSource` diagnostic, if you would prefer, but we don't have anything so special-cased as "also check for `_d`"; I've never seen this before.

**Comment by daltairwalter** [comment]
We get our python from ActiveState and my understanding is that their business is to build opensource projects like python and tcl and sell the builds - so this is likely to be common.  I think they follow the python standard on this.  Searching on Google for the "_d" returns a lot of results.

**Comment by jakebailey** [comment]
@brettcannon I haven't really been able to find much on this, other than a hint that debug builds of Python may look instead for `_d` suffixed compiled modules instead of normal ones (but that doesn't seem like the example here). Are you aware of something like this?

**Comment by brettcannon** [comment]
If you run a debug build of Python it will look for `_d.pyd` files in older versions of Python since the symbols are different from a release build of Python. I believe this is no longer the case with newer versions of Python.

@daltairwalter what version of Python are you using?

**Comment by daltairwalter** [comment]
python 3.5.  We are in the process of updating our python.

**Comment by jakebailey** [comment]
It seems like special casing would be needed to support this, but I'm a bit confused why someone would be distributing a build like this at all, if it requires a completely different set of compiled modules (that would then be incompatible with regular modules).

**Comment by daltairwalter** [comment]
From my perspective, you have the question backwards.  Why isn't everyone distributing builds like this?  It seems like there are two alternatives:
1. only build a release python - this is great until one is developing projects with embedded python and creating lots of python extensions.  There are a lot of hacks required to build in this partial release / partial debug format and it is very, very expensive.
2. Build both a full python debug toolchain as well as the full release python.  From what I have heard, there are a lot of difficulties building standard libraries like numpy,scipy etc. in debug.  I think this is why more people are not building debug.

We definitely like this debug distribution and it was one of the big selling points for us using ActiveState instead of building everything ourselves.

**Comment by jakebailey** [comment]
I'm not intending to really advise on something like that; this is just the first time I've seen this in my time working on tooling for Python.

To do this, we'd need to make our import resolver debug-build aware (it already calls out to get paths), at the cost of making that call (which for nearly all users won't change the behavior).

If it's true that more recent versions of Python don't even need this (and instead can interchange modules), then it seems like this special case is only going to get more rare over time.

**Comment by daltairwalter** [comment]
I agree with your statement about newer versions of python.  If the future of python is to remove the _d this doesn't make sense.  My understanding from the few comments I saw while I was looking was that the _d was added in python 3.5 - not removed.

As far as expense goes, searching for an extra file seems very cheap by comparison to outputting the warning.

**Comment by jakebailey** [comment]
Well, the cost is the maintenance of the import resolver being more complicated, but I was mainly referring to the cost of calling out to Python, which is not very fast on Windows. We could try to lump that in with some other call, of course. Just thinking aloud.

**Comment by erictraut** [comment]
If this behavior is deprecated after Python 3.5, then it's probably not worth adding. If it is meant to work in later versions of Python, then it might be worth exploring a solution.

@jakebailey, why do you think we'd need to call out to Python? I was thinking that we could simply look for both ".pyd" and "_d.pyd" files and consider the import resolved if either are present.

**Comment by jakebailey** [comment]
If only a debug build of Python looks for these, then allowing them to resolve may produce a condition where we don't complain about an import that would fail at runtime.

I suppose the "it's rare" argument could also imply that it's not such a big deal to just add a check for `_d` to the compiled code import checks and have the false negative, with the assumption that most never even have these files.

I'd have to find some documentation about this; all I've been able to find are random StackOverflow posts referring to it and no official source, but I'm probably just searching for the wrong thing.

**Comment by brettcannon** [comment]
Unfortunately I misspoke as only the stable ABI doesn't require rebuilds to work with debug builds of Python: https://docs.python.org/3/whatsnew/3.8.html#debug-build-uses-the-same-abi-as-release-build.

See https://github.com/python/cpython/blob/794430700defb913512f871b701a888aa730de81/Lib/importlib/_bootstrap_external.py#L815 and https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Python/dynload_win.c#L15-L19 showing that `_d.pyd` is still being supported.

`importlib.machinery.EXTENSION_SUFFIXES` will list what suffixes are supported for extension modules.

**Comment by jakebailey** [comment]
Thanks for doing the investigation.

It at least looks like this is limited to Windows, but given we don't actually do platform-specific import resolution (e.g. we'll accept a `.pyd` on Linux, or `.so` on Windows, etc), maybe we just add it as a valid suffix and call it good.

**Comment by daltairwalter** [comment]
I tried the importlib.machinery.EXTENSION_SUFFIXES in a few places and I see things like:
Anaconda:
['.cp37-win_amd64.pyd', '.pyd']

ActiveState release
['.cp35-win_amd64.pyd', '.pyd']

ActiveState debug
['_d.cp35-win_amd64.pyd', '_d.pyd']

The ActiveState 3.8.10 debug version appears to be keeping to these - at least with the '_d.cp38-win_amd64.pyd'

All of the prebuilt pyds that come with the installation seem to have the longer '_d.cp38-win_amd64.pyd' suffix

**Comment by heejaechang** [comment]
since it only happens with debug bits and we have workaround by disabling missing import source diagnostic, put it in p2 with enhancement.

asking is making "resolveNativeImportEx" to understand "_d" postfix special casing for debug native modules.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
